### PR TITLE
fix dtype handling in 2 optimizers and 1 layer

### PR DESCRIPTION
**Pull request by bstriner**
Update `Adam` and `RMSprop` so dtype of momentum variables matches dtype of corresponding parameter. Shouldn't affect most people unless you're using a mix of dtypes. Would be nice if `K.zeros_like` used the parameter's dtype but that might break a lot of other things. May be worth updating the other optimizers but those are the two I use the most.

Update `Embedding` layer to have a `output_dtype` parameter that is passed to `add_weight`. Useful if you need mixed precision or if you have integer embeddings (`trainable=False`).

Cheers

### CODE REVIEW

**Comment by bstriner** [comment]
I need to write handling for "[float32]_ref" to just be "float32". Issue causes failure in tensorflow. Passes in theano.

No easy way in keras to make zeros like something else, because tensorflow dtype might report "float32_ref".

Maybe a tensorflow function to strip the "ref" when using K.dtype(x)?

**Comment by fchollet** [comment]
I believe that conceptually the `dtype` argument of a layer should be the dtype of the weight matrices of the layer, not that of its input. Input dtype can be specified in input spec for validation checks, if needed.

So I'm looking at why we had dtype `int32` in the first place. It may be that we can just remove that, in which case your argument `output_dtype` would just be `dtype`.

File: keras/layers/embeddings.py [hunk 1]
```diff
@@ -55,6 +55,7 @@ class Embedding(Layer):
           This argument is required if you are going to connect
           `Flatten` then `Dense` layers upstream
           (without it, the shape of the dense outputs cannot be computed).
+      output_dtype: dtype of embedding weights. Default (`None`) uses `K.floatx()`.
```
Comment (fchollet) [inline keras/layers/embeddings.py]: As far as I can tell you can remove this argument and do the following:
- remove `kwargs['dtype'] = 'int32'`
- use `dtype=self.output_dtype` in `add_weight`

**Comment by fchollet** [comment]
We may need to update the implementation of `K.dtype` in the TF backend to be consistent with the other backends. I believe there is a TF dtype method that will return the "vanilla" dtype name in every case.

**Comment by bstriner** [comment]
@fchollet please let me know your thoughts on this update to `K.dtype(x)`. There is an extra parameter `K.dtype(x, base_dtype=True)`.

The idea is that `True` should always return a standard keras dtype and `False` can return something specific to the backend. Most things should work with `True` unless you're doing something backend specific.

```
if base_dtype:
  return x.dtype.base_dtype.name
else:
  return x.dtype.name
```

The alternative would be no parameter, and `K.dtype` always returns the base type.

Cheers

**Comment by ahundt** [comment]
`base_dtype` should probably be a string, like `'backend'` or `'python'` or `'tensorflow'`

**Comment by bstriner** [comment]
@ahundt in that case would the parameter be better named something like `mode`? Are there more than 2? If there are a lot of modes, separate functions might be better.

Current implementation of `dtype` doesn't remove `_ref` in tensorflow. PR defaults to remove the `_ref` but has a parameter if you want to not remove it.

I'm not sure whether keeping the `_ref` is useful in some cases. It is the current behavior and is not the default in the PR, so might change things for some people. Does it seem handy to have a function that gives the actual raw dtype of the backend? Should it be the same function or a separate function?

Just brainstorming, but convert `K.dtype` to always get the keras dtype, then add something like `K.raw_dtype` if you want the raw backend dtype for some reason. If `raw_dtype` isn't that necessary, demote it to `keras_contrib`.

**Comment by ahundt** [comment]
Defaulting to the keras dtype sounds appropriate, even if all that's implemented are the cases you first described as a true false case I was thinking a string would be clearer and prevent breaking changes in the future.

**Comment by fchollet** [comment]
The backend API is a shared interface across different platforms. It doesn't need to support TF-only functionality: if people want TF-only functionality in their code, they can use the TF API for that.

> The alternative would be no parameter, and K.dtype always returns the base type.

I would go with that solution.

**Comment by bstriner** [comment]
Fixes are in. `K.dtype` not returning the basetype in tensorflow was a bug. Change to existing behavior but should make things more consistent. No explanation for the hardcoded `int32` but removing it doesn't seem to have an impact.

**Comment by fchollet** [comment]
> No explanation for the hardcoded int32 but removing it doesn't seem to have an impact.

Pretty sure it was legacy code.

**Pull request merged**